src/clarity-angular/input-trap/input-trap.directive.ts

```diff
@@ -0,0 +1,47 @@
+import { Directive, HostListener, ElementRef, OnInit } from "@angular/core";
+
+@Directive({
+    selector: "[clrInputTrap]"
+})
+export class InputTrapDirective implements OnInit {
+
+    static TAB_KEY = 9;
+
+    // tslint:disable-next-line:max-line-length
+    // reference: https://github.com/jkup/focusable
+    static FOCUSABLE_INPUTS = "a[href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), button:not([disabled])";
+
+    firstInput: HTMLElement;
+    lastInput: HTMLElement;
+
+    constructor(public elementRef: ElementRef) { }
+
+    @HostListener("document:keydown", ["$event"])
+    onkeydown(event: any) {
```

Comment:
@victormejia: Actually, the `focusin` pattern wouldn't need you to track focusable elements at all anymore. What you do is check the target of the `focusin` event:
- If it's in the modal, you do nothing.
- If it's outside of the modal, then you re-focus *on the modal itself*. You don't need to focus necessarily on a focusable element, focusing on the modal element is good enough.

This would make it support naturally tabindex, ignore hidden elements, handle dynamic elements without killing performances, ...